xds/src/test/java/io/grpc/xds/LrsLoadBalancerTest.java

```diff
@@ -0,0 +1,325 @@
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.LoadBalancer.ResolvedAddresses;
+import io.grpc.LoadBalancer.Subchannel;
+import io.grpc.LoadBalancer.SubchannelPicker;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.Status;
+import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.xds.ClientLoadCounter.LoadRecordingStreamTracerFactory;
+import io.grpc.xds.ClientLoadCounter.LoadRecordingSubchannelPicker;
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.LrsLoadBalancerProvider.LrsConfig;
+import java.net.SocketAddress;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+/**
+ * Unit tests for {@link LrsLoadBalancer}.
+ */
+@RunWith(JUnit4.class)
+public class LrsLoadBalancerTest {
+  @Rule
+  public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  private static final String CLUSTER_NAME = "cluster-foo.googleapis.com";
+  private static final String EDS_SERVICE_NAME = "cluster-foo:service-blade";
+  private static final String LRS_SERVER_NAME = "trafficdirector.googleapis.com";
+  private static final Locality TEST_LOCALITY =
+      new Locality("test-region", "test-zone", "test-subzone");
+
+  private final ClientLoadCounter counter = new ClientLoadCounter();
+  private final LoadRecorder loadRecorder = new LoadRecorder();
+  private final Queue<LoadBalancer> childBalancers = new ArrayDeque<>();
+
+  @Mock
+  private Helper helper;
+
+  private LrsLoadBalancer loadBalancer;
+
+  @Before
+  public void setUp() {
+    loadBalancer = new LrsLoadBalancer(helper);
+  }
+
+  @After
+  public void tearDown() {
+    loadBalancer.shutdown();
+    assertThat(loadRecorder.recording).isFalse();
+  }
+
+  @Test
+  public void subchannelPickerInterceptedWithLoadRecording() {
+    List<EquivalentAddressGroup> backendAddrs = createResolvedBackendAddresses(2);
+    deliverResolvedAddresses(backendAddrs, "round_robin");
+    assertThat(loadRecorder.recording).isTrue();
+    ArgumentCaptor<SubchannelPicker> pickerCaptor = ArgumentCaptor.forClass(null);
+    verify(helper).updateBalancingState(eq(ConnectivityState.READY), pickerCaptor.capture());
+    SubchannelPicker picker = pickerCaptor.getValue();
+    assertThat(picker).isInstanceOf(LoadRecordingSubchannelPicker.class);
+    PickResult result = picker.pickSubchannel(mock(PickSubchannelArgs.class));
+    ClientStreamTracer.Factory tracerFactory = result.getStreamTracerFactory();
+    assertThat(((LoadRecordingStreamTracerFactory) tracerFactory).getCounter())
+        .isSameInstanceAs(counter);
+    assertThat(result.getSubchannel().getAllAddresses()).isEqualTo(backendAddrs);
```

Comment:
This line is just testing test implementation and nothing else is tested. Use mock `subchannel` with `childHelper.updateBalancingState()` as test input, and verify `result.getSubchannel()` is expected.

Reply:
Deleted. We only care about the picker, how subchannel are created with respect to addresses are completely up to child balancer/helper's implementation. Not interested by `LrsLoadBalancer`'s behavior.